src/test/java/com/microsoft/sqlserver/jdbc/preparedStatement/BatchExecutionWithBulkCopyTest.java

```diff
@@ -0,0 +1,359 @@
+        method.setAccessible(true);
+
+        ArrayList<String> valueList = (ArrayList<String>) method.invoke(pstmt, false);
+        ArrayList<String> valueListExpected = new ArrayList<String>();
+        valueListExpected.add("1");
+        valueListExpected.add("2");
+        valueListExpected.add("'?'");
+        valueListExpected.add("?");
+
+        for (int i = 0; i < valueListExpected.size(); i++) {
+            assertEquals(valueList.get(i), valueListExpected.get(i));
+        }
+    }
+    
+    @Test
+    public void testAllcolumns() throws Exception {
+        Field f1 = SQLServerConnection.class.getDeclaredField("isAzureDW");
+        f1.setAccessible(true);
+        f1.set(connection, true);
+        
+        String valid = "INSERT INTO " + tableName + " values "
+                + "("
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + ")";
+        
+        SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) connection.prepareStatement(valid);
+        SQLServerStatement stmt = (SQLServerStatement) connection.createStatement();
+        
+        Timestamp myTimestamp = new Timestamp(114550L);
+        
+        Date d = new Date(114550L);
+        
+        pstmt.setInt(1, 1234);
+        pstmt.setBoolean(2, false);
+        pstmt.setString(3, "a");
+        pstmt.setDate(4, d);
+        pstmt.setDateTime(5, myTimestamp);
+        pstmt.setFloat(6, (float) 123.45);
+        pstmt.setString(7, "b");
+        pstmt.setString(8, "varc");
+        pstmt.setString(9, "''");
+        pstmt.addBatch();
+        
+        pstmt.executeBatch();
+        
+        ResultSet rs = stmt.executeQuery("SELECT * FROM " + tableName);
+        
+        Object[] expected = new Object[9];
+        
+        expected[0] = 1234;
+        expected[1] = false;
+        expected[2] = "a";
+        expected[3] = d;
+        expected[4] = myTimestamp;
+        expected[5] = 123.45;
+        expected[6] = "b";
+        expected[7] = "varc";
+        expected[8] = "''";
+        
+        rs.next();
+        for (int i=0; i < expected.length; i++) {
+            assertEquals(rs.getObject(i + 1).toString(), expected[i].toString());
+        }
+    }
+
+    @Test
+    public void testMixColumns() throws Exception {
+        Field f1 = SQLServerConnection.class.getDeclaredField("isAzureDW");
+        f1.setAccessible(true);
+        f1.set(connection, true);
+        
+        String valid = "INSERT INTO " + tableName + " (c1, c3, c5, c8) values "
+                + "("
+                + "?, "
+                + "?, "
+                + "?, "
+                + "?, "
+                + ")";
+        
+        SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) connection.prepareStatement(valid);
```

Comment:
local `pstmt` and `stmt` are never closed.

Reply:
Sure. I made them all re-use the member variable stmt and pstmt instead.